hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/storage/row/HoodieRowParquetWriteSupport.java

```diff
@@ -73,6 +189,180 @@ public void add(UTF8String recordKey) {
         bloomFilterWriteSupport.addKey(recordKey));
   }
 
+  @FunctionalInterface
+  private interface ValueWriter {
+    void write(SpecializedGetters row, int ordinal);
+  }
+
+  private void consumeMessage(Runnable writer) {
+    recordConsumer.startMessage();
+    writer.run();
+    recordConsumer.endMessage();
+  }
+
+  private void consumeGroup(Runnable writer) {
+    recordConsumer.startGroup();
+    writer.run();
+    recordConsumer.endGroup();
+  }
+
+  private void consumeField(String field, int index, Runnable writer) {
+    recordConsumer.startField(field, index);
+    writer.run();
+    recordConsumer.endField(field, index);
+  }
+
+  private void writeFields(InternalRow row, StructType schema, ValueWriter[] fieldWriters) {
+    for (int i = 0; i < row.numFields(); i++) {
+      int index = i;
+      if (!row.isNullAt(i)) {
+        StructField field = schema.fields()[i];
+        consumeField(field.name(), index, () -> fieldWriters[index].write(row, index));
+      }
```

Comment:
nit: reminder to remove `index`; `field = schema.fields()[i]` can also be inlined on L220

Reply:
Updated